### DISCUSSION TITLE
Best way to investigating oscillation errors inside custom components ?

### DISCUSSION THREAD

**Discussion by jongena** [hneemann/Digital#1017] [Q&A] [answered]
Perhaps my question is not sufficiently clear. I reformulate it slightly. 

A large circuit contains several rather large and complex custom components with many in- and outputs.
Testing the full circuit results (not unexpectedly) in oscillation errors.
Digital indicates that they are located inside the custom components.  

When I open the indicated custom components, nothing is indicated in the opened component, although the
software apparantly knows what is going on. Is there a way to see where the error occurs in these opened custom components ?

The only way out - in as far as my understanding of the software is correct -  is to include the "problematic custom components" explicitly in the final circuit ? It would simplify life a lot if that could be avoided. Is this possible ? 
Or is there another solution that makes an easy debugging possible of  "problematic" custom circuits ?

**Comment by hneemann** [comment, marked as answer]
Hmm! If the oscillation also affects the containing circuit, it should not be a problem, since you can see which wires are affected. If the oscillation only occurs in the embedded circuit, however, the same input pattern must be generated in the circuit as the containing circuit does. And this is really not always so easy.
But often the cause of the oscillation can be seen by simply looking at the circuit. There must be an unbuffered feedback. Often there are not so many.

**Reply by jongena** [reply]
Many thanks for the feedback !

**Reply by prlaba** [reply]
One cause of oscillation errors I discovered is improper initialization of all memory components in a circuit.  I had a flip-flop in a feedback path to a ROM which wasn't being reset at power on.  Once I reset it with a POR signal the oscillation error disappeared.

**Reply by jongena** [reply]
Again thanks for the feedback. I implemented a POR signal and got now rid of the oscillations.

**Reply by prlaba** [reply]
Great!  So much easier finding these issues in simulation rather than in the actual circuits.
